infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/provision/InstallerConfigApplier.java

```diff
@@ -0,0 +1,131 @@
+ * otherwise changing configuration has no effect.
+ *
+ * <p>This class performs following changes to environment:
+ * <br> - adds environment variable to {@link Container containers} that are required by installers or agents.
+ * <br> - adds environment variables which are specified in properties of installer configuration.
+ * The environment property contains environment variables in the following format: "env1=value1,env2=value2,...";
+ * <br> - performs all required changes that are needed for exposing installers' servers.
+ *
+ * @author Sergii Leshchenko
+ */
+public class InstallerConfigApplier {
+    private static final Logger LOG = getLogger(InstallerConfigApplier.class);
+
+    private static final String ENVIRONMENT_PROPERTY = "environment";
+
+    private final InstallerRegistry installerRegistry;
+    private final String            cheServerEndpoint;
+
+    @Inject
+    public InstallerConfigApplier(InstallerRegistry installerRegistry,
+                                  @Named("che.infra.openshift.che_server_endpoint") String cheServerEndpoint) {
+        this.installerRegistry = installerRegistry;
+        this.cheServerEndpoint = cheServerEndpoint;
+    }
+
+    public void apply(RuntimeIdentity identity,
+                      Environment envConfig,
+                      OpenShiftEnvironment environment) throws InfrastructureException {
+        for (Pod pod : environment.getPods().values()) {
+            String podName = pod.getMetadata().getName();
+            for (Container container : pod.getSpec().getContainers()) {
+                String containerName = container.getName();
+                String machineName = podName + "/" + containerName;
+                MachineConfig machineConf = envConfig.getMachines().get(machineName);
+
+                List<String> installers = machineConf.getInstallers();
+                Map<String, ServerConfig> name2Server = new HashMap<>();
+                for (String installerId : installers) {
+                    Installer installer = getInstaller(installerId);
+                    provisionEnv(container, installer.getProperties());
+                    name2Server.putAll(installer.getServers());
+                }
+
+                ServerExposeUtils.expose(machineName,
+                                         container,
+                                         name2Server,
+                                         environment,
+                                         "agents");
+
+                // CHE_API is used by installers for agent binary downloading
+                container.getEnv().removeIf(env -> "CHE_API".equals(env.getName()));
+                container.getEnv().add(new EnvVar("CHE_API", cheServerEndpoint, null));
+
+                // WORKSPACE_ID is required only by workspace agent
+                if (installers.contains("org.eclipse.che.ws-agent")) {
+                    container.getEnv().removeIf(env -> "CHE_WORKSPACE_ID".equals(env.getName()));
```

Comment:
I agree with you. It would be nice to have common provision mechanism and then what infrastructure will have to implement is just applying prepared changes to its objects.
But now it will not be done in the scope of this issue.
It requires changes in the model because now it doesn't have environment variables anymore. @garagatyi WDYT will we implement this?